File: problems-extend/1501-1600/1506_find-root-of-n-ary-tree/solution_child_set.py
from typing import List, Optional


# Judge-provided types (not editable here; the judge assembles their
# definitions into every submission):
#   Node:      .val int, .children list[Node]; the tree arrives as its node list


class Solution:
    def findRoot(self, tree: List[Node]) -> Optional[Node]:
        # Indegree zero: every node except the root appears exactly once as
        # someone's child. Collect all the nodes, then discard every node
        # seen as a child — the one survivor is the root.
        survivors = set(tree)
        for node in tree:
            for child in node.children:
                survivors.discard(child)
        for node in survivors:
            return node
        return None

File: problems-originals/1501-1600/1506_find-root-of-n-ary-tree/solution_child_set.go
func findRoot(tree []*Node) *Node {
	// Indegree zero: every node except the root appears exactly once as
	// someone's child. Collect all the nodes, then discard every node seen
	// as a child — the one survivor is the root.
	survivors := make(map[*Node]bool, len(tree))
	for _, node := range tree {
		survivors[node] = true
	}
	for _, node := range tree {
		for _, child := range node.Children {
			delete(survivors, child)
		}
	}
	for node := range survivors {
		return node
	}
	return nil
}
